### PR TITLE
Port oauth 2 1 post

### PR DESCRIPTION
**Pull request by mooreds**
This is the oauth 2.1 blog post ported over to the expert advice section.

### CODE REVIEW

File: site/learn/expert-advice/oauth/whats-new-with-oauth-2-1.md [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+
+That bears repeating. Nothing new will be added. This is an explicit design goal of OAuth 2.1.
+
+This article assumes you are a developer or have analogous technical experience and are familiar with OAuth and the terms used in the various corresponding standards. If you’d like an introduction to OAuth and why you’d consider using it, [Wikipedia is a good place to start](https://en.wikipedia.org/wiki/OAuth). This article covers proposed changes to OAuth that might affect you if you are either using OAuth in your application or if you build software which implements OAuth.
```
Comment (dpolstra) [inline site/learn/expert-advice/oauth/whats-new-with-oauth-2-1.md:16]: Maybe link to https://fusionauth.io/learn/expert-advice/oauth/oauth-v1-signed-requests/? Maybe: "If you’d like an introduction to OAuth and why you’d consider using it, check out this Wikipedia overview and our expert advice article on V1 signed requests"

File: site/learn/expert-advice/oauth/whats-new-with-oauth-2-1.md [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+
+Many of the details are drawn from the [OAuth 2.0 Security Best Current Practices](https://tools.ietf.org/html/draft-ietf-oauth-security-topics-14) document. In the name of security best practices, some of the more problematic grants will be removed.
+
+At the end of the day, the goal of OAuth 2.1 is to have a single document explaining how to best implement and use OAuth, as both a client and an authorization server. No longer will developers be required to hunt across multiple RFCs and standards documents to understand how a particular behavior should be implemented or used.
+
+## What does OAuth 2.1 mean to me?
+
+If you use OAuth in your application, don’t panic. 
+
+As mentioned above, the discussion process around this specification is ongoing. A draft was posted to the IETF mailing list in mid July and was adopted by the OAuth working group in late July. As of the time of writing, the draft is being reviewed and fine tuned.
```
Comment (dpolstra) [inline site/learn/expert-advice/oauth/whats-new-with-oauth-2-1.md:38]: July 2020

File: site/learn/expert-advice/oauth/whats-new-with-oauth-2-1.md [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+
+Here's a suggested description pulled from the [ongoing mailing list discussion](https://mailarchive.ietf.org/arch/msg/oauth/Ne4Q9erPP7SpC5051sSy6XnLDv0/): 
+
+> By design, [OAuth 2.1] does not introduce any new features to what already exists in the OAuth 2.0 specifications being replaced.
+
+Many of the details are drawn from the [OAuth 2.0 Security Best Current Practices](https://tools.ietf.org/html/draft-ietf-oauth-security-topics-14) document. In the name of security best practices, some of the more problematic grants will be removed.
+
+At the end of the day, the goal of OAuth 2.1 is to have a single document explaining how to best implement and use OAuth, as both a client and an authorization server. No longer will developers be required to hunt across multiple RFCs and standards documents to understand how a particular behavior should be implemented or used.
+
+## What does OAuth 2.1 mean to me?
```
Comment (dpolstra) [inline site/learn/expert-advice/oauth/whats-new-with-oauth-2-1.md:34]: How will OAuth 2.1 affect you?

File: site/learn/expert-advice/oauth/whats-new-with-oauth-2-1.md [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+
+You may ask, when will it be available? The short answer is "no one knows". 
+
+The long answer is "truly, no one knows. It does seem like we're getting along in the process, though." 
+
+Even after OAuth 2.1 is released, it will likely be some time before it is widely implemented. Omitted grants may be supported with, with nags or warnings, forever. The exact changes to the specification are still being discussed and the publishing of an RFC is further in the future. Finally, when OAuth 2.1 is released, you can continue to use an OAuth 2.0 server if that meets your needs. 
+
+That said, when OAuth 2.1 is released the largest impact on anyone using OAuth for authentication or authorization in their applications will be adapting to the removal of two OAuth 2.0 specified grants: the Implicit grant and the Resource Owner Password Credentials grant. 
+
+## What is changing?
```
Comment (dpolstra) [inline site/learn/expert-advice/oauth/whats-new-with-oauth-2-1.md:48]: What is changing from OAuth 2.0 to OAuth 2.1?

File: site/learn/expert-advice/oauth/whats-new-with-oauth-2-1.md [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+
+The [Proof Key for Code Exchange (PKCE) RFC](https://tools.ietf.org/html/rfc7636) was published in 2015 and extends the Authorization Code grant to protect from an attack if part of the authorization flow happens over a non TLS connection. For example, this occurs if there's communication between components of a native application. 
+
+This attack could also happen if TLS has a vulnerability or if a user's router firmware has been compromised and is spoofing DNS or downgrading TLS to HTTP. PKCE requires an additional one-time code to be generated and sent to the OAuth server. This code validates the request has not been intercepted or modified. 
+
+The OAuth 2.1 draft specification requires the PKCE challenge be used with every Authorization Code grant, protecting against the authorization code being hijacked by an attacker and being used to gain a token.
+
+### Redirect URIs must be compared using exact string matching
```
Comment (dpolstra) [inline site/learn/expert-advice/oauth/whats-new-with-oauth-2-1.md:82]: OAuth 2.1 change: Redirect URIs must be compared using exact string matching

File: site/learn/expert-advice/oauth/whats-new-with-oauth-2-1.md [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+
+> Refresh tokens must either be sender-constrained or one-time use as per Section 4.12.2 of [OAuth 2.0 Security Best Current Practices](https://tools.ietf.org/html/draft-ietf-oauth-security-topics-15)
+
+Whew, that's a lot. Let's examine each of these in turn. But before we do, let's define a few terms used in the rest of this article.
+
+* A `client` is a piece of code that the user is interacting with; browsers, native apps or single-page applications are all clients. 
+* An `OAuth server` implements OAuth specifications. It has or can obtain information about which resources are available to clients. In the RFCs this application is called an Authorization Server. This is also known as an Identity Provider. Most users call it "the place I sign in". 
+* An `application server` doesn’t have any authentication functionality but instead delegates login requests to an OAuth server. It has an id which allows the OAuth server to identify it. 
+
+### The Authorization Code grant and PKCE
```
Comment (dpolstra) [inline site/learn/expert-advice/oauth/whats-new-with-oauth-2-1.md:70]: OAuth 2.1 change: The Authorization Code grant and PKCE

File: site/learn/expert-advice/oauth/whats-new-with-oauth-2-1.md [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+
+An additional use case for a wild card redirect URI occurs when the final destination page needs dynamic parameters, like `trackingparam=123&specialoffer=abc`. These are usually appended to the redirect URI before the OAuth process begins. However, without wild cards, a URL with dynamic parameters won't match any of the configured redirect URIs, so the redirect fails.
+
+However, allowing wildcard matching for the redirect URI is a security risk. If the redirect URI matching is flexible, an attacker could send a user to an open redirect server controlled by them, and then on to a malicious destination. OWASP also discusses the [perils of such open redirect servers](https://cheatsheetseries.owasp.org/cheatsheets/Unvalidated_Redirects_and_Forwards_Cheat_Sheet.html#dangerous-url-redirects). While this attack requires compromising the request in some fashion, using exact matching for redirect URIs eliminates this risk because the redirect URI is always a known value.
+
+### The Implicit grant is removed
```
Comment (dpolstra) [inline site/learn/expert-advice/oauth/whats-new-with-oauth-2-1.md:98]: OAuth 2.1 change: The Implicit grant is removed

File: site/learn/expert-advice/oauth/whats-new-with-oauth-2-1.md [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+
+Here's an [article about the least insecure way to use the Implicit grant for an SPA](https://fusionauth.io/learn/expert-advice/authentication/spa/oauth-implicit-grant-jwts-cookies). In this situation, the access token is provided to the SPA as an HttpOnly cookie. In the end it basically recreates the Authorization Code grant, illustrating how the Implicit grant simply can't be made secure.
+
+The OAuth 2.1 draft specification omits the Implicit grant. This means that any software implementing OAuth 2.1 won't have to implement this grant. If you use this grant in your application, you’ll have to replace it with a different one if you want to be compliant with OAuth 2.1. 
+
+### The Resource Owner Password Credentials grant is removed
```
Comment (dpolstra) [inline site/learn/expert-advice/oauth/whats-new-with-oauth-2-1.md:110]: OAuth 2.1 change: The Resource Owner Password Credentials grant is removed

File: site/learn/expert-advice/oauth/whats-new-with-oauth-2-1.md [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+
+This grant is often used for native mobile applications. While this grant made it easier to migrate to OAuth with minimal application changes, it doesn't follow the typical delegation pattern. After all, the application server has full access to the credentials of the user, exactly what OAuth is designed to avoid. No longer can you leave the work of securing users' credentials and data to the OAuth server. With this grant, you must ensure that your application backend is just as secure.
+
+If you have a mobile application using this grant, you can either update the client to use an Authorization Code grant using PKCE or keep using your OAuth 2.0 compliant system.
+
+### No bearer tokens in the query string
```
Comment (dpolstra) [inline site/learn/expert-advice/oauth/whats-new-with-oauth-2-1.md:120]: OAuth 2.1 change: No bearer tokens in the query string

File: site/learn/expert-advice/oauth/whats-new-with-oauth-2-1.md [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+
+Bearer tokens, also known as access tokens, allow access to protected resources and therefore must be secured. They are called bearer tokens because merely possessing them gives you access. They're like a set of house keys. 
+
+These days, most tokens are JWTs. Clients store them securely and then use them to make API calls to the application server. The application server then uses the token to ensure the client calling the API has appropriate permissions. When first defined in [RFC 6750](https://tools.ietf.org/html/rfc6750), tokens were allowed in headers, POST bodies or query strings. The OAuth 2.1 draft prohibits sending a bearer token in a query string. 
+
+A query string and, more generally, any string in a URL, is never private. JavaScript executing on a page can access it. A URL or components thereof may be recorded in server log files, caches, or browser history. In general, if you want to pass information privately over the internet, don't use anything in a URL. Instead, use TLS and put the sensitive information in a POST body or HTTP header.
+
+### Limiting refresh tokens
```
Comment (dpolstra) [inline site/learn/expert-advice/oauth/whats-new-with-oauth-2-1.md:130]: OAuth 2.1 change: Limiting refresh tokens

File: site/learn/expert-advice/oauth/whats-new-with-oauth-2-1.md [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+
+A query string and, more generally, any string in a URL, is never private. JavaScript executing on a page can access it. A URL or components thereof may be recorded in server log files, caches, or browser history. In general, if you want to pass information privately over the internet, don't use anything in a URL. Instead, use TLS and put the sensitive information in a POST body or HTTP header.
+
+### Limiting refresh tokens 
+
+> Refresh tokens must either be sender-constrained or one-time use as per Section 4.12.2 of [OAuth 2.0 Security Best Current Practices](https://tools.ietf.org/html/draft-ietf-oauth-security-topics-14)
+
+[Refresh tokens](https://tools.ietf.org/html/rfc6749#section-6) allow a client to retrieve new access tokens without reauthentication by the resource owner. If you need access to a resource for longer than an access token lives, or if you need infrequent access, refresh tokens can help. One example of where a refresh token might be used is when a user is logged into their email for weeks or months. Every time an access token expires, the client can present a refresh token and get a new access token.
```
Comment (dpolstra) [inline site/learn/expert-advice/oauth/whats-new-with-oauth-2-1.md:134]: Can we link to https://fusionauth.io/docs/v1/tech/oauth/tokens/ in this section?

File: site/learn/expert-advice/oauth/whats-new-with-oauth-2-1.md [hunk 12]
```diff
@@ -0,0 +1,170 @@
+---
+layout: advice
+title: What is new in the OAuth 2.1?
+description: A draft of the OAuth 2.1 specification was recently released. What's coming down the pike?
+image: advice/oauth-device-authorization-article.png
+author: Dan Moore
+category: OAuth
+date: 2020-10-20
+dateModified: 2020-10-20
```
Comment (dpolstra) [inline site/learn/expert-advice/oauth/whats-new-with-oauth-2-1.md:9]: Make sure we have the correct redirect in place. This post gets a lot of traffic.

File: site/learn/expert-advice/oauth/whats-new-with-oauth-2-1.md [hunk 13]
```diff
@@ -0,0 +1,170 @@
+---
+layout: advice
+title: What is new in the OAuth 2.1?
```
Comment (dpolstra) [inline site/learn/expert-advice/oauth/whats-new-with-oauth-2-1.md:3]: Maybe change the title to "Changes to expect when OAuth 2.1 launches"

File: site/learn/expert-advice/oauth/whats-new-with-oauth-2-1.md (hunk 12 above)
Comment (mooreds) [reply site/learn/expert-advice/oauth/whats-new-with-oauth-2-1.md:9]: I wasn't going to redirect. This is similar to what we did here: https://github.com/FusionAuth/fusionauth-site/pull/199/

So was planning to not even add a link on the blog post to this expert advice, just let it be present in the expert advice area and bulk that up some.

File: site/learn/expert-advice/oauth/whats-new-with-oauth-2-1.md (hunk 11 above)
Comment (mooreds) [reply site/learn/expert-advice/oauth/whats-new-with-oauth-2-1.md:134]: Sure thing.

File: site/learn/expert-advice/oauth/whats-new-with-oauth-2-1.md (hunk 12 above)
Comment (dpolstra) [reply site/learn/expert-advice/oauth/whats-new-with-oauth-2-1.md:9]: My concern with doing it this way is being penalized for duplicate content

File: site/learn/expert-advice/oauth/whats-new-with-oauth-2-1.md (hunk 1 above)
Comment (mooreds) [reply site/learn/expert-advice/oauth/whats-new-with-oauth-2-1.md:16]: Yes, doesn't really fit, unfortunately. When Brian releases the OAuth guide, we should point to that.

File: site/learn/expert-advice/oauth/whats-new-with-oauth-2-1.md (hunk 13 above)
Comment (mooreds) [reply site/learn/expert-advice/oauth/whats-new-with-oauth-2-1.md:3]: I went with "Differences between OAuth 2 and OAuth 2.1"

File: site/learn/expert-advice/oauth/whats-new-with-oauth-2-1.md (hunk 12 above)
Comment (mooreds) [reply site/learn/expert-advice/oauth/whats-new-with-oauth-2-1.md]: I understand. I thought that if we changed ~10% of the content we wouldn't have to worry about that.

Comment (mooreds) [reply site/learn/expert-advice/oauth/whats-new-with-oauth-2-1.md]:
```
old: 3136 words  2468 78% common  134 4% deleted  534 17% changed
new: 3264 words  2468 75% common  212 6% inserted  584 17% changed
```

**Pull request merged**